geth/common/queued_tx.go

```diff
@@ -0,0 +1,166 @@
+package common
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+)
+
+const (
+	// MessageIDKey is a key for message ID
+	// This ID is required to track from which chat a given send transaction request is coming.
+	MessageIDKey = contextKey("message_id")
+)
+
+type contextKey string // in order to make sure that our ctx key does not collide with keys from other packages
+
+var (
+	//ErrQueuedTxInProgress - error transaction in progress
+	ErrQueuedTxInProgress = errors.New("transaction is in progress")
+	//ErrQueuedTxDiscarded - error transaction discarded
+	ErrQueuedTxDiscarded = errors.New("transaction has been discarded")
+)
+
+// QueuedTx holds enough information to complete the queued transaction.
+type QueuedTx struct {
+	id         QueuedTxID
+	hash       common.Hash
+	ctx        context.Context
+	args       SendTxArgs
+	inProgress bool // true if transaction is being sent
+	done       chan error
+	doneOnce   sync.Once
+	err        error
+	sync.RWMutex
+}
+
+// NewQueuedTx QueuedTx constructor.
+func NewQueuedTx(ctx context.Context, id QueuedTxID, args SendTxArgs) *QueuedTx {
+	return &QueuedTx{
+		id:   id,
+		ctx:  ctx,
+		args: args,
+		done: make(chan error, 1),
+	}
+}
+
+// ID gets queued transaction ID.
+func (tx *QueuedTx) ID() QueuedTxID {
+	tx.RLock()
+	defer tx.RUnlock()
+
+	return tx.id
+}
+
+// Hash gets queued transaction hash.
+func (tx *QueuedTx) Hash() common.Hash {
+	tx.RLock()
+	defer tx.RUnlock()
+
+	return tx.hash
+}
+
+// MessageID gets message ID from ctx.
+func (tx *QueuedTx) MessageID() string {
+	tx.RLock()
+	defer tx.RUnlock()
+
+	if tx.ctx == nil {
+		return ""
+	}
+
+	if messageID, ok := tx.ctx.Value(MessageIDKey).(string); ok {
+		return messageID
+	}
+
+	return ""
+}
+
+// Args gets queued transaction args.
+func (tx *QueuedTx) Args() SendTxArgs {
+	tx.RLock()
+	defer tx.RUnlock()
+
+	return tx.args
+}
+
+// UpdateGasPrice updates gas price if not set.
+func (tx *QueuedTx) UpdateGasPrice(gasGetter func() (*hexutil.Big, error)) (*hexutil.Big, error) {
```

Comment:
What's a point in using a getter as an argument instead of passing an actual value? If `gasGetter` requires an internet connection, you will block a `QueuedTx` for hundreds of milliseconds or even more. Anything else like checking `QueuedTx` state will need to wait for finishing that call.

Reply:
I've just save old logic: if `tx.args.GasPrice == nil` we do make a call for gasPrice in other case we can return gasPrice immediately. It makes sence i think.

Reply:
Not really sure you read my comment carefully :P You fundamentally changed when a call for the gas price is done. Previously, the call was not sent when the lock was acquired.